### ISSUE TITLE
java 13 text blocks syntax is not supported

### ISSUE THREAD

**Issue by joxerTMD** [checkstyle/checkstyle#7103]
Java 13 has been released and now it is possible to write multiline string literals
```
String multiLineStr = 
"""
   SELECT * FROM USERS
   WHERE USER_ID = '1'
""";
```
Unfortunately checkstyle fails to parse it now and i have to make a difficult choice between turning off checkstyle in my project and using not  the latest version of java.

Do you have any plans to support it in the nearest future?

**Comment by strkkk** [comment]
@joxerTMD 
Multiline strings are just a [preview feature](https://openjdk.java.net/jeps/355) so I guess it should become an official part of language first.

**Comment by romani** [comment]
@joxerTMD, yes it is hard choice, as such syntax feature is so long awaited.

We will definitely do support for it, but we will wait a bit to let java seal syntax of it.

Meanwhile I can only suggest to use https://checkstyle.org/config_filefilters.html#BeforeExecutionExclusionFileFilter to not let checkstyle try to parse it.

I make issue as approved, but I keep right to not merge proposed fix till final release of multiline strings in kava if grammar change on our side will result in some regression.

**Comment by estekhin** [comment]
It means that checkstyle can not be used at all in any project that moves to Java 13 and uses text blocks at least until Java 14, may be until some later Java version if 14 will not include text blocks as non-preview feature.

The argument that the text block grammar could change is understood, but could you consider supporting it as "preview" feature which could be enabled by the checkstyle user on their discretion?

**Comment by romani** [comment]
Yes we are ok to accept fix, if doesn't  require massive update in all other Checks, as AST structure might change.
PR is welcome, we will merge quickly only if we minimal regression on others. All depends on implementation in PR.

**Comment by pbludov** [comment]
We have a number of checks for string literals, for example https://checkstyle.org/config_coding.html#MultipleStringLiterals
They will need to be reviewed once this issue will be done.

**Comment by romani** [comment]
It might be good to run our usual diff report but with all Checks (+ sevntu checks) activated.

**Comment by jordanjennings** [comment]
With the release of Java 14 and the second preview of text blocks appearing to be the final syntax per [JEP-378](https://openjdk.java.net/jeps/378), it would be fantastic to add this support. Looks like some good work underway as of January to do it. Any chance of it getting integrated soon?

**Comment by romani** [comment]
I think we will make support of it till end of August. Watch for this issue, you will be notified when something is done.

**Comment by nmancus1** [comment]
@esilkensen @pbludov if this looks good, I should be able to submit a PR this week:

```
➜  text-blocks-grammar cat TestClass.java 
public class TestClass {
    static {
        String doubleQuotes = """
                ""
                """;
        String oneDoubleQuote = """
                "
                """;
        String empty = """
                """;
        String oneSingleQuote = """
                '
                """;
        String escape = """
                <html>\u000D\u000A\n
                    <body>\u000D\u000A\n
                        <p>Hello, world</p>\u000D\u000A\n
                    </body>\u000D\u000A\n
                </html>\u000D\u000A
                """;
        String testMoreEscapes = """
                fun with\n
                whitespace\t\r
                and other escapes \"""
                """;
        String sqlQuery = """
                SELECT `EMP_ID`, `LAST_NAME` FROM `EMPLOYEE_TB`
                WHERE `CITY` = 'INDIANAPOLIS'
                ORDER BY `EMP_ID`, `LAST_NAME`;
                """;
        String concat = """
                The quick brown fox""" + "  \n" + """
                jumps over the lazy dog
                """;
    }
}
➜  text-blocks-grammar java -jar ~/IdeaProjects/checkstyle/target/checkstyle-8.34-SNAPSHOT-all.jar -T TestClass.java
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|   `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:7]
|--IDENT -> TestClass [1:13]
`--OBJBLOCK -> OBJBLOCK [1:23]
    |--LCURLY -> { [1:23]
    |--STATIC_INIT -> STATIC_INIT [2:4]
    |   `--SLIST -> { [2:11]
    |       |--VARIABLE_DEF -> VARIABLE_DEF [3:8]
    |       |   |--MODIFIERS -> MODIFIERS [3:8]
    |       |   |--TYPE -> TYPE [3:8]
    |       |   |   `--IDENT -> String [3:8]
    |       |   |--IDENT -> doubleQuotes [3:15]
    |       |   `--ASSIGN -> = [3:28]
    |       |       `--EXPR -> EXPR [3:30]
    |       |           `--TEXT_BLOCK_LITERAL_BEGIN -> """ [3:30]
    |       |               |--TEXT_BLOCK_CONTENT -> \n                ""\n                 [3:30]
    |       |               `--TEXT_BLOCK_LITERAL_END -> """ [4:38]
    |       |--SEMI -> ; [4:38]
    |       |--VARIABLE_DEF -> VARIABLE_DEF [5:8]
    |       |   |--MODIFIERS -> MODIFIERS [5:8]
    |       |   |--TYPE -> TYPE [5:8]
    |       |   |   `--IDENT -> String [5:8]
    |       |   |--IDENT -> oneDoubleQuote [5:15]
    |       |   `--ASSIGN -> = [5:30]
    |       |       `--EXPR -> EXPR [5:32]
    |       |           `--TEXT_BLOCK_LITERAL_BEGIN -> """ [5:32]
    |       |               |--TEXT_BLOCK_CONTENT -> \n                "\n                 [5:32]
    |       |               `--TEXT_BLOCK_LITERAL_END -> """ [6:37]
    |       |--SEMI -> ; [6:37]
    |       |--VARIABLE_DEF -> VARIABLE_DEF [7:8]
    |       |   |--MODIFIERS -> MODIFIERS [7:8]
    |       |   |--TYPE -> TYPE [7:8]
    |       |   |   `--IDENT -> String [7:8]
    |       |   |--IDENT -> empty [7:15]
    |       |   `--ASSIGN -> = [7:21]
    |       |       `--EXPR -> EXPR [7:23]
    |       |           `--TEXT_BLOCK_LITERAL_BEGIN -> """ [7:23]
    |       |               |--TEXT_BLOCK_CONTENT -> \n                 [7:23]
    |       |               `--TEXT_BLOCK_LITERAL_END -> """ [8:19]
    |       |--SEMI -> ; [8:19]
    |       |--VARIABLE_DEF -> VARIABLE_DEF [9:8]
    |       |   |--MODIFIERS -> MODIFIERS [9:8]
    |       |   |--TYPE -> TYPE [9:8]
    |       |   |   `--IDENT -> String [9:8]
    |       |   |--IDENT -> oneSingleQuote [9:15]
    |       |   `--ASSIGN -> = [9:30]
    |       |       `--EXPR -> EXPR [9:32]
    |       |           `--TEXT_BLOCK_LITERAL_BEGIN -> """ [9:32]
    |       |               |--TEXT_BLOCK_CONTENT -> \n                '\n                 [9:32]
    |       |               `--TEXT_BLOCK_LITERAL_END -> """ [11:19]
    |       |--SEMI -> ; [11:19]
    |       |--VARIABLE_DEF -> VARIABLE_DEF [12:8]
    |       |   |--MODIFIERS -> MODIFIERS [12:8]
    |       |   |--TYPE -> TYPE [12:8]
    |       |   |   `--IDENT -> String [12:8]
    |       |   |--IDENT -> escape [12:15]
    |       |   `--ASSIGN -> = [12:22]
    |       |       `--EXPR -> EXPR [12:24]
    |       |           `--TEXT_BLOCK_LITERAL_BEGIN -> """ [12:24]
    |       |               |--TEXT_BLOCK_CONTENT -> \n                <html>\u000D\u000A\n\n                    <body>\u000D\u000A\n\n                        <p>Hello, world</p>\u000D\u000A\n\n                    </body>\u000D\u000A\n\n                </html>\u000D\u000A\n                 [12:24]
    |       |               `--TEXT_BLOCK_LITERAL_END -> """ [18:19]
    |       |--SEMI -> ; [18:19]
    |       |--VARIABLE_DEF -> VARIABLE_DEF [19:8]
    |       |   |--MODIFIERS -> MODIFIERS [19:8]
    |       |   |--TYPE -> TYPE [19:8]
    |       |   |   `--IDENT -> String [19:8]
    |       |   |--IDENT -> testMoreEscapes [19:15]
    |       |   `--ASSIGN -> = [19:31]
    |       |       `--EXPR -> EXPR [19:33]
    |       |           `--TEXT_BLOCK_LITERAL_BEGIN -> """ [19:33]
    |       |               |--TEXT_BLOCK_CONTENT -> \n                fun with\n\n                whitespace\t\r\n                and other escapes \"""\n                 [19:33]
    |       |               `--TEXT_BLOCK_LITERAL_END -> """ [22:58]
    |       |--SEMI -> ; [22:58]
    |       |--VARIABLE_DEF -> VARIABLE_DEF [23:8]
    |       |   |--MODIFIERS -> MODIFIERS [23:8]
    |       |   |--TYPE -> TYPE [23:8]
    |       |   |   `--IDENT -> String [23:8]
    |       |   |--IDENT -> sqlQuery [23:15]
    |       |   `--ASSIGN -> = [23:24]
    |       |       `--EXPR -> EXPR [23:26]
    |       |           `--TEXT_BLOCK_LITERAL_BEGIN -> """ [23:26]
    |       |               |--TEXT_BLOCK_CONTENT -> \n                SELECT `EMP_ID`, `LAST_NAME` FROM `EMPLOYEE_TB`\n                WHERE `CITY` = 'INDIANAPOLIS'\n                ORDER BY `EMP_ID`, `LAST_NAME`;\n                 [23:26]
    |       |               `--TEXT_BLOCK_LITERAL_END -> """ [27:19]
    |       |--SEMI -> ; [27:19]
    |       |--VARIABLE_DEF -> VARIABLE_DEF [28:8]
    |       |   |--MODIFIERS -> MODIFIERS [28:8]
    |       |   |--TYPE -> TYPE [28:8]
    |       |   |   `--IDENT -> String [28:8]
    |       |   |--IDENT -> concat [28:15]
    |       |   `--ASSIGN -> = [28:22]
    |       |       `--EXPR -> EXPR [29:48]
    |       |           `--PLUS -> + [29:48]
    |       |               |--PLUS -> + [29:39]
    |       |               |   |--TEXT_BLOCK_LITERAL_BEGIN -> """ [28:24]
    |       |               |   |   |--TEXT_BLOCK_CONTENT -> \n                The quick brown fox [28:24]
    |       |               |   |   `--TEXT_BLOCK_LITERAL_END -> """ [29:38]
    |       |               |   `--STRING_LITERAL -> "  \n" [29:41]
    |       |               `--TEXT_BLOCK_LITERAL_BEGIN -> """ [29:50]
    |       |                   |--TEXT_BLOCK_CONTENT -> \n                jumps over the lazy dog\n                 [29:50]
    |       |                   `--TEXT_BLOCK_LITERAL_END -> """ [31:19]
    |       |--SEMI -> ; [31:19]
    |       `--RCURLY -> } [32:4]
    `--RCURLY -> } [33:0]

```

**Comment by esilkensen** [comment]
Thanks @nmancus1, that looks good to me

**Comment by pbludov** [comment]
One thought is that the Java compiler removes whitespace at the beginning of the each line. AST can be similar to AST for Javadoc:
```
`--EXPR -> EXPR [23:26]
   `--TEXT_BLOCK_LITERAL_BEGIN -> """ [23:26]
       |--TEXT_BLOCK_CONTENT -> \n                SELECT `EMP_ID`, `LAST_NAME` FROM `EMPLOYEE_TB`\n                WHERE `CITY` = 'INDIANAPOLIS'\n                ORDER BY `EMP_ID`, `LAST_NAME`;\n                 [23:26]
         |--NL -> '\n'
         |--WHITESPACE ->  '                '
         |--TEXT -> SELECT `EMP_ID`, `LAST_NAME` FROM `EMPLOYEE_TB`
         |--NL -> '\n'
         |--WHITESPACE ->  '                '
         |--TEXT -> WHERE `CITY` = 'INDIANAPOLIS'
         |--NL -> '\n'
         |--WHITESPACE ->  '                '
         |--TEXT -> ORDER BY `EMP_ID`, `LAST_NAME`;
         '--NL -> '\n'
       `--TEXT_BLOCK_LITERAL_END -> """ [27:19]
```

But I'm not sure it's a good idea. For this block
```java
        String sqlQuery = """
                SELECT `EMP_ID`, `LAST_NAME` FROM `EMPLOYEE_TB`
                    WHERE `CITY` = 'INDIANAPOLIS'
                    ORDER BY `EMP_ID`, `LAST_NAME`;
                """;
```
The text will be
```
SELECT `EMP_ID`, `LAST_NAME` FROM `EMPLOYEE_TB`
    WHERE `CITY` = 'INDIANAPOLIS'
    ORDER BY `EMP_ID`, `LAST_NAME`;
```
So it's not obvious where the whitespace ends and the text begins.

I agree with this proposal. It can be improved in the future, if necessary.

**Comment by nmancus1** [comment]
@pbludov What do you think about this:
```
➜  text-blocks-grammar cat TestClass.java
public class TestClass {
    String sqlQuery = """
                SELECT `EMP_ID`, `LAST_NAME` FROM `EMPLOYEE_TB`
                WHERE `CITY` = 'INDIANAPOLIS'
                ORDER BY `EMP_ID`, `LAST_NAME`;
                """;

}
➜  text-blocks-grammar java -jar ~/IdeaProjects/checkstyle/target/checkstyle-8.34-SNAPSHOT-all.jar -T TestClass.java
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|   `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:7]
|--IDENT -> TestClass [1:13]
`--OBJBLOCK -> OBJBLOCK [1:23]
    |--LCURLY -> { [1:23]
    |--VARIABLE_DEF -> VARIABLE_DEF [2:4]
    |   |--MODIFIERS -> MODIFIERS [2:4]
    |   |--TYPE -> TYPE [2:4]
    |   |   `--IDENT -> String [2:4]
    |   |--IDENT -> sqlQuery [2:11]
    |   |--ASSIGN -> = [2:20]
    |   |   `--EXPR -> EXPR [2:22]
    |   |       `--TEXT_BLOCK_LITERAL_BEGIN -> """ [2:22]
    |   |           |--TEXT_BLOCK_CONTENT ->                 SELECT `EMP_ID`, `LAST_NAME` FROM `EMPLOYEE_TB` [3:0]
    |   |           |--NEWLINE -> \n [3:63]
    |   |           |--TEXT_BLOCK_CONTENT ->                 WHERE `CITY` = 'INDIANAPOLIS' [4:0]
    |   |           |--NEWLINE -> \n [4:45]
    |   |           |--TEXT_BLOCK_CONTENT ->                 ORDER BY `EMP_ID`, `LAST_NAME`; [5:0]
    |   |           |--NEWLINE -> \n [5:47]
    |   |           |--TEXT_BLOCK_CONTENT ->                  [6:0]
    |   |           `--TEXT_BLOCK_LITERAL_END -> """ [6:16]
    |   `--SEMI -> ; [6:19]
    `--RCURLY -> } [8:0]

```

**Comment by pbludov** [comment]
> @pbludov What do you think about this:

I don't see how this option is good. For text blocks, we may need two views: the original (raw) one, as in the file: with all whitespace, and the one that the compiler produces after removing the initial spaces. Let's leave the raw version, and use some utility method to get the trimmed version if need.

**Comment by nmancus1** [comment]
I haven't removed any white space; this is just as it is in the file, line by line. I believe this approach is superior since you can clearly see what text is on which line, and what column that the text ends on. This approach also allows users to easily discern between a true newline, and one that is typed in the text block. After seeing your idea above, it became clear to me that it would be backwards to take a nicely formatted text block, and basically make it into a very long one line string in the AST. Also, when printing the AST in a soft wrapped terminal, the long lines wrap around and make the tree difficult to read.

**Comment by esilkensen** [comment]
I think the AST in https://github.com/checkstyle/checkstyle/issues/7103#issuecomment-645028044 matches what I would it expect it to be after reading the spec, i.e. "A text block consists of zero or more content characters, enclosed by opening and closing delimiters." It makes me think of those 3 AST nodes:
- `TEXT_BLOCK_LITERAL_BEGIN` "The opening delimiter is a sequence of three double quote characters (""") followed by zero or more white spaces followed by a line terminator."
- `TEXT_BLOCK_LITERAL_END` "The closing delimiter is a sequence of three double quote characters."
- `TEXT_BLOCK_CONTENT` "The content begins at the first character after the line terminator of the opening delimiter. The content ends at the last character before the first double quote of the closing delimiter."

>I believe this approach is superior since you can clearly see what text is on which line. This way also allows users to easily discern between a true newline, and one that is typed in the text block. After seeing your idea above, it became clear to me that it would be backwards to take a nicely formatted text block (in the file), and basically make it into a very long one line string in the AST. Also, when printing the AST in a soft wrapped terminal, the long lines wrap around and make the tree difficult to read:

I think these are good thoughts, but that it's worth asking too, what is the usability of this AST from the perspective of someone writing or configuring a check in terms of it?

Or, is it more/less difficult for code analyzing one of these AST's to have the content node split into multiple parts by `NEWLINE` nodes?

I think it's more important to design for that than the human readability of the printed out AST.

So I actually think it makes sense "to take a nicely formatted text block (in the file), and basically make it into a very long one line string in the AST" -- presumably someone wishing to recover the nicely formatted text block could just grab the contents of a `TEXT_BLOCK_CONTENT` and print it; and it seems simpler to me for the AST to be at that level of abstraction, i.e. what is the _content_ of this text block as a single value.

**Comment by esilkensen** [comment]
>I believe this approach is superior since you can clearly see what text is on which line. This way also allows users to easily discern between a true newline, and one that is typed in the text block. 

I'm not sure I follow here: what do you mean by true newline vs. one typed in a text block?

>After seeing your idea above, it became clear to me that it would be backwards to take a nicely formatted text block (in the file), and basically make it into a very long one line string in the AST. Also, when printing the AST in a soft wrapped terminal, the long lines wrap around and make the tree difficult to read:

If the only reason to add the extra `NEWLINE` structure to the AST is visualizing it / human readability of the printed tree, then it seems better to me to keep the simpler tree that you described in https://github.com/checkstyle/checkstyle/issues/7103#issuecomment-645028044.

On the other hand if there were checks or other code that would benefit from having the text block contents themselves already parsed for newlines, then I could see that having them is beneficial.

But my feeling right now is that it actually makes sense to take the nicely formatted text block and turn it into a very long "one line" string, i.e. the raw value of the string, in the AST. This matches what I would expect from reading the spec:

 "A text block consists of zero or more content characters, enclosed by opening and closing delimiters." 
- `TEXT_BLOCK_LITERAL_BEGIN` "The opening delimiter is a sequence of three double quote characters (""") followed by zero or more white spaces followed by a line terminator."
- `TEXT_BLOCK_LITERAL_END` "The closing delimiter is a sequence of three double quote characters."
- `TEXT_BLOCK_CONTENT` "The content begins at the first character after the line terminator of the opening delimiter. The content ends at the last character before the first double quote of the closing delimiter."

Then like @pbludov said there could be a utility function that extracts the processed text block value if necessary from the raw `TEXT_BLOCK_CONTENT`.

**Comment by nmancus1** [comment]
>I'm not sure I follow here: what do you mean by true newline vs. one typed in a text block?

Since we can use escaped characters such as `\n` (as in the `String escape` example above), when you read example 1, you see: `<html>\u000D\u000A\n\n`.  In example 2, you only see `<html>\u000D\u000A\n` that is intended to be the "content" of the text block, not the real control characters. the "real" newline is the newline that is after the first TEXT_BLOCK_CONTENT node.  

>if there were checks or other code that would benefit from having the text block contents themselves already parsed for newlines

This is a good point, and should really be the deciding factor. After a quick review of our checks, it looks like the extra nodes will only complicate things, especially checks that are children of `TreeWalker`.

>If the only reason to add the extra NEWLINE structure to the AST is visualizing it / human readability of the printed tree

Yes, this was my main intention initially.  It stemmed from seeing @pbludov 's example above.  But upon implementing this view, it became clear that it is critical to discern the true control characters from those that are "content" to be able to keep track of the row/ line.  I don't think anyone noticed, but in my [first example](https://github.com/checkstyle/checkstyle/issues/7103#issuecomment-645028044), as well as the [first PR for this issue](https://github.com/checkstyle/checkstyle/issues/7103#issuecomment-645028044) the line numbers are incorrect.  Since in example 1, the entire content of the text block is being parsed as one long string, the lexer is never picking up the true control characters and calling `newline() ` when it should. 

So, it is important to parse these correctly, regardless of the approach we take.  If we don't want to print them in the AST, that's understandable, since it would make the tree significantly longer in some cases and could complicate checks.  I can just make the modified NEWLINE protected, and incorporate it into the lexer rule for TEXT_BLOCK_CONTENT so that `newline()` is called when appropriate.

In summation, I agree with @esilkensen  and @pbludov that this could unnecessarily complicate checks.  I will move forward with the approach taken in https://github.com/checkstyle/checkstyle/issues/7103#issuecomment-645028044 this week, and amend my current PR to reflect this decision.

**Comment by rnveach** [comment]
Fix is merged
